src/java.base/share/classes/jdk/internal/math/DoubleToDecimal.java

```diff
@@ -149,71 +156,42 @@ public static String toString(double v) {
      * {@link #toString(double) rendered} and the resulting string were then
-     * {@link Appendable#append(CharSequence) appended} to {@code app}.
      *
-     * @param v the {@code double} whose rendering is appended.
-     * @param app the {@link Appendable} to append to.
+     * @param str the String byte array to append to
+     * @param index the index into str
+     * @param v the {@code double} whose rendering is into str.
      * @throws IOException If an I/O error occurs
      */
-    public static Appendable appendTo(double v, Appendable app)
-            throws IOException {
-        return new DoubleToDecimal(false).appendDecimalTo(v, app);
-    }
+    public int putDecimal(byte[] str, int index, double v) {
+        assert 0 <= index && index <= length(str) - MAX_CHARS : "Trusted caller missed bounds check";
 
-    private String toDecimalString(double v) {
-        return switch (toDecimal(v, null)) {
-            case NON_SPECIAL -> charsToString();
-            case PLUS_ZERO -> "0.0";
-            case MINUS_ZERO -> "-0.0";
-            case PLUS_INF -> "Infinity";
-            case MINUS_INF -> "-Infinity";
-            default -> "NaN";
-        };
-    }
-
-    private Appendable appendDecimalTo(double v, Appendable app)
-            throws IOException {
-        switch (toDecimal(v, null)) {
-            case NON_SPECIAL:
-                char[] chars = new char[index + 1];
-                for (int i = 0; i < chars.length; ++i) {
-                    chars[i] = (char) bytes[i];
-                }
-                if (app instanceof StringBuilder builder) {
-                    return builder.append(chars);
-                }
-                if (app instanceof StringBuffer buffer) {
-                    return buffer.append(chars);
-                }
-                for (char c : chars) {
-                    app.append(c);
-                }
-                return app;
-            case PLUS_ZERO: return app.append("0.0");
-            case MINUS_ZERO: return app.append("-0.0");
-            case PLUS_INF: return app.append("Infinity");
-            case MINUS_INF: return app.append("-Infinity");
-            default: return app.append("NaN");
+        int pair = toDecimal(str, index, v, null);
+        int type = pair & 0xFF00;
+        if (type == NON_SPECIAL) {
+            return index + (pair & 0xFF);
         }
+        return putSpecial(str, index, type);
     }
 
     /*
-     * Returns
+     * Returns size in the lower byte, type in the high byte, where type is
      *     PLUS_ZERO       iff v is 0.0
      *     MINUS_ZERO      iff v is -0.0
      *     PLUS_INF        iff v is POSITIVE_INFINITY
      *     MINUS_INF       iff v is NEGATIVE_INFINITY
      *     NAN             iff v is NaN
+     *     otherwise NON_SPECIAL
      */
-    private int toDecimal(double v, FormattedFPDecimal fd) {
+    private int toDecimal(byte[] str, int index, double v, FormattedFPDecimal fd) {
```

Comment:
This `toDecimal` returns packed int with info but the other returns the new index; this is extremely confusing for future readers, especially that they have very similar signatures (difference in signature is in the middle). I recommend making this return a `long` so when it's downcasted to an int, it's just the size (like String concat's coderIndex)

Reply:
The maximum length of float to decimal is 15, and the maximum length of double to decimal is 24, so one byte is enough. If we use the long return type, we have to add `cast int to long` and `cast long to int` to the code, which is redundant. It will also confuse people, thinking that float to decimal and double to decimal will have a large length, just like the size of string concat, which requires a complete int.